packages/connect-extension-protocol/src/index.ts

```diff
@@ -1,143 +1,98 @@
- * as recieved by the `ExtensionProvider`.
- *
- * @remarks The browser wraps the message putting it in {@link data}
- */
-export interface ExtensionMessage {
-  data: ExtensionMessageData
+interface ToWebpageError {
+  type: ToWebpageMessageType.Error
+  payload: string
 }
-export interface ExtensionMessageData {
-  /** origin is used to determine which side sent the message **/
-  origin: "content-script"
-  /** message is telling the `ExtensionProvider` the port has been closed **/
-  disconnect?: boolean
-  /** message is the message from the manager to be forwarded to the app **/
-  message?: MessageFromManager
+
+export type ToWebpageBody = ToWebpageDisconnect | ToWebpageRpc | ToWebpageError
+
+export type ToWebpage = {
+  header: ToWebpageHeader
+  body: ToWebpageBody
 }
 
-/**
- * ExtensionListenHandler is a message handler type for receiving
- * `ProviderMessage` messages from the \@substrate/connect `ExtensionProvider`
- * in the extension.
- */
-export type ExtensionListenHandler = (message: ProviderMessage) => void
-
-/**
- * extension provides strongly typed convenience wrappers around
- * the `window.postMessage` and `window.addEventListener` APIs used for
- * message passing on the extension side of communication.
- */
 export const extension = {
-  /** send a message from the extension to the app **/
-  send: (message: ExtensionMessageData): void => {
+  send: (message: ToWebpage): void => {
     window.postMessage(message, "*")
   },
-  /**
-   * Listen to messages from the `ExtensionProvider` in the app sent to
-   * the extension.
-   */
-  listen: (handler: ExtensionListenHandler): void => {
+  listen: (handler: EventHandler<ToExtension>): void => {
     window.addEventListener("message", handler)
   },
 }
 
-/**
- * ProviderMessage represents messages sent via `window.postMessage` from
- * `ExtensionProvider` to `ExtensionMessageRouter` as received by the extension.
- *
- * @remarks The browser wraps the message putting it in {@link data}
- */
-export interface ProviderMessage {
-  data: ProviderMessageData
+export interface ToExtensionHeader {
+  origin: "extension-provider"
+  providerId: number
 }
 
-export interface ProviderMessageData {
-  /** origin is used to determine which side sent the message **/
-  origin: "extension-provider"
-  /** The name of the app to be used for display purposes in the extension UI **/
-  appName: string
-  /** The uniqueId for extension multiplexing **/
-  chainId: number
-  /** The name of the blockchain network the app is talking to **/
-  chainName: string
-  /** What action the `ExtensionMessageRouter` should take **/
-  action: "forward" | "connect" | "disconnect"
-  /** The message the `ExtensionMessageRouter` should forward to the background **/
-  message?: MessageToManager
+export enum ToExtensionMessageType {
+  Connect,
+  Disconnect,
+  Spec,
+  Rpc,
+}
+
+interface ToExtensionConnect {
+  type: ToExtensionMessageType.Connect
```

Comment:
What is this compiled to? A string containing "Connect"?

Reply:
In the way that the enum has been defined it currently [compiles to a number](https://www.typescriptlang.org/docs/handbook/enums.html), we could make it compile to a string... but why does it matter? The type system won't allow using the underlying compiled value, so I don't see why that's important.

Reply:
It matters because the communication is ultimately done in JavaScript. If the webpage <-> extension protocol relies on the implementation details of TypeScript then you couldn't write code that talks to the extension without using TypeScript yourself.
I would strongly prefer strings here.

Reply:
Changing the order of the elements in the enum, or removing some elements of the enum, would be a breaking change to the protocol.

Reply:
> It matters because the communication is ultimately done in JavaScript. If the webpage <-> extension protocol relies on the implementation details of TypeScript then you couldn't write code that talks to the extension without using TypeScript yourself.

That's a good point, I will change that.

Reply:
done! Thanks for pointing that out!